# Data Science Pandas

## Tasks Today:

0) <b>Pre-Work</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Numpy Random Sampling

1) <b>Pandas</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Importing <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Tabular Data Structures <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - from_dict() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - read_csv() <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) <b>In-Class Exercise #1</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Accessing Data <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Indexing <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - df.loc <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - keys() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Slicing a DataFrame <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Built-In Methods <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - head() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - tail() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - shape <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - describe() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - sort_values() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - .columns <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) <b>In-Class Exercise #2</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; g) Filtration <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Conditionals <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Subsetting <br>
 &nbsp;&nbsp;&nbsp;&nbsp; h) Column Transformations <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Generating a New Column w/Data <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - User Defined Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; i) Aggregations <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - groupby() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Type of groupby() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - mean() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - groupby() w/Multiple Columns <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - drop_duplicates() <br>

## Numpy Random Sampling

In [2]:
import numpy as np

# A single call generates a single random number
print('Here is a random number: %s' % np.random.uniform())

# You can also pass some bounds
print('Here is a random number between 0 and 1 Million: %s' % np.random.uniform(0, 1e6))

# You can also generate a bunch of random numbers all at once
print('Here are 3 random numbers between 0 and 1 Million: %s' % np.random.uniform(0, 1e6, 3))

# Even matrices and tensors with weird shapes
print('Here is a 3x3 matrix of random numbers between 0 and 1 Million: \n %s' % np.random.uniform(0, 1e6, (3, 3)))

# instead of float values, generating random integers
print('Using Random Integer: %s' % np.random.randint(0, 10, 4))

Here is a random number: 0.07364686248430796
Here is a random number between 0 and 1 Million: 882060.4308056345
Here are 3 random numbers between 0 and 1 Million: [827060.4012427   12500.77770067  45149.02547802]
Here is a 3x3 matrix of random numbers between 0 and 1 Million: 
 [[677923.32922485 240676.51203298  63153.65320375]
 [958741.97784717 601104.04740907 875860.38152397]
 [837633.94712365 911040.96820532 766954.87752546]]
Using Random Integer: [0 6 9 4]


## Pandas <br>

<p>Pandas is a flexible data analysis library built on top of NumPy that is excellent for working with tabular data. It is currently the de-facto standard for Python-based data analysis, and fluency in Pandas will do wonders for your productivity and frankly your resume. It is one of the fastest ways of getting from zero to answer in existence. </p>

<ul>
    <li>Pandas is a Python module, written in C. The Pandas module is a high performance, highly efficient, and high level data analysis library. It allows us to work with large sets of data called dataframes.</li>
    <li>Series is a one-dimensional labeled array capable of holding data of any type (integer, string, float, python objects, etc.)</li>
    <li>Dataframe = Spreadsheet (has column headers, index, etc.)</li>
</ul>

### Importing

In [3]:
import pandas as pd

# always use pd, standard for data science

### Tabular data structures <br>
<p>The central object of study in Pandas is the DataFrame, which is a tabular data structure with rows and columns like an excel spreadsheet. The first point of discussion is the creation of dataframes both from native Python dictionaries, and text files through the Pandas I/O system.</p>

In [4]:
names = ['Alice',
         'Bob',
         'James',
         'Beth', 
         'John', 
         'Sally',
         'Richard', 
         'Lauren',
         'Brandon', 
         'Sabrina']

ages = np.random.randint(18,35,len(names)) #random Ages
my_people = {
    'names': names,
    'ages': ages
}

print(my_people)

{'names': ['Alice', 'Bob', 'James', 'Beth', 'John', 'Sally', 'Richard', 'Lauren', 'Brandon', 'Sabrina'], 'ages': array([22, 20, 32, 27, 32, 33, 24, 19, 31, 21])}


##### from_dict()

<p>Let's convert our not-so-useful-for-analysis dict into a Pandas dataframe. We can use the from_dict function to do this easily using Pandas:</p>

In [5]:
data = pd.DataFrame.from_dict(my_people)
print('Resulting type of data: {}'.format(type(data)))
data

Resulting type of data: <class 'pandas.core.frame.DataFrame'>


names  ages
0    Alice    22
1      Bob    20
2    James    32
3     Beth    27
4     John    32
5    Sally    33
6  Richard    24
7   Lauren    19
8  Brandon    31
9  Sabrina    21

##### read_csv()

In [ ]:
marathon = pd.read_csv('../files/boston_marathon2017_edited.csv', sep =',')
marathon

### In-Class Exercise #1 - Read in Boston Red Sox Hitting Data <br>
<p>Use the pandas read_csv() method to read in the statistics from the two files yesterday.</p>

In [6]:
boston2017 = pd.read_csv('../files/redsox_2017_hitting.txt', sep =',')
boston2018 = pd.read_csv('../files/redsox_2018_hitting.txt', sep =',')

boston2018

Rk Pos                Name  Age    G   PA   AB    R    H  2B  ...    OBP  \
0    1   C          Sandy Leon   29   89  288  265   30   47  12  ...  0.232   
1    2  1B      Mitch Moreland   32  124  459  404   57   99  23  ...  0.325   
2    3  2B       Eduardo Nunez   31  127  502  480   56  127  23  ...  0.289   
3    4  SS     Xander Bogaerts   25  136  580  513   72  148  45  ...  0.360   
4    5  3B       Rafael Devers   21  121  490  450   59  108  24  ...  0.298   
5    6  LF   Andrew Benintendi   23  148  661  579  103  168  41  ...  0.366   
6    7  CF  Jackie Bradley Jr.   28  144  535  474   76  111  33  ...  0.314   
7    8  RF        Mookie Betts   25  136  614  520  129  180  47  ...  0.438   
8    9  DH       J.D. Martinez   30  150  649  569  111  188  37  ...  0.402   
9   10  MI          Brock Holt   30  109  367  321   41   89  18  ...  0.362   
10  11   C   Christian Vazquez   27   80  269  251   24   52  10  ...  0.257   
11  12  UT       Blake Swihart   26   82  207  192   28   44  10  ...  0.285   
12  13  UT      Hanley Ramirez   34   44  195  177   25   45   7  ...  0.313   
13  14  1B        Steve Pearce   35   50  165  136   19   38   8  ...  0.394   
14  15  2B         Ian Kinsler   36   37  143  132   17   32   6  ...  0.294   
15  16  SS         Tzu-Wei Lin   24   37   73   65   15   16   6  ...  0.329   
16  17  UT          Sam Travis   24   19   38   36    5    8   3  ...  0.263   
17  18  UT    Brandon Phillips   37    9   27   23    4    3   0  ...  0.259   
18  19  2B      Dustin Pedroia   34    3   13   11    1    1   0  ...  0.231   
19  20   C          Dan Butler   31    2    7    6    0    1   0  ...  0.143   

      SLG    OPS  OPS+   TB  GDP  HBP  SH  SF  IBB  
0   0.279  0.511    37   74    6    4   3   1    0  
1   0.433  0.758   102  175   12    0   0   5    2  
2   0.388  0.677    81  186   17    2   1   3    0  
3   0.522  0.883   135  268   14    6   0   6    4  
4   0.433  0.731    94  195    9    0   0   2    6  
5   0.465  0.830   123  269    9    2   2   7    1  
6   0.403  0.717    92  191    6   11   0   4    3  
7   0.640  1.078   186  333    5    8   0   5    8  
8   0.629  1.031   173  358   19    4   0   7   11  
9   0.411  0.774   109  132    7    7   0   2    2  
10  0.283  0.540    46   71    5    4   1   0    1  
11  0.328  0.613    65   63    4    0   0   0    0  
12  0.395  0.708    90   70    9    2   0   2    0  
13  0.507  0.901   141   69    1    5   0   2    0  
14  0.311  0.604    64   41    5    0   0   1    0  
15  0.415  0.744   100   27    0    0   0   0    0  
16  0.389  0.652    73   14    1    0   0   0    0  
17  0.261  0.520    42    6    1    0   0   0    0  
18  0.091  0.322    -7    1    0    0   0   0    0  
19  0.167  0.310   -17    1    0    0   0   1    0  

[20 rows x 28 columns]

### Accessing Data <br>

##### Indexing

<p>You can directly select a column of a dataframe just like you would a dict. The result is a Pandas 'Series' object.</p>

In [17]:
# Even though they are more complex series objects, they still support the behavior of the underlying NumPy arrays
data_ages = data['ages']
print(type(data_ages))

print(data_ages)
 # key first, row index second
print(data_ages[3])

print(data['names'][0])


<class 'pandas.core.series.Series'>
0    29
1    26
2    29
3    32
4    19
5    32
6    23
7    31
8    26
9    31
Name: ages, dtype: int32
32
Alice


##### df.loc

<p>Along the horizontal dimension, rows of Pandas DataFrames are Row objects. You will notice there is a third column present in the DataFrame - this is the $\textit{index}$. It is automatically generated as a row number, but can be reassigned to a column of your choice using the DataFrame.set_index(colname) method. We can use it to access particular Pandas $\textit{rows}$, which are also Series objects:</p>

In [19]:
my_row = data.loc[0]
print(my_row)

print(data['names'][0])
# index after assinging the variable with .loc

# indexing without defining variable


names    Alice
ages        29
Name: 0, dtype: object
Alice


##### keys()

In [20]:
# A series can support dict-like features when it's used as a DataFrame row. 
my_row.keys()

Index(['names', 'ages'], dtype='object')

##### Slicing a DataFrame

In [22]:
 # slicing allowed
print(data[2:5])

# Slice for one item
print(data[2:3])
# The above syntax will work, but be aware, passing as single int will be interpreted as a column key
# df[2] # single indexing not allowed

   names  ages
2  James    29
3   Beth    32
4   John    19
   names  ages
2  James    29


### Built-In Methods <br>

<p>These are methods that are frequently used when using Pandas to make your life easier. It is possible to spend a whole week simply exploring the built-in functions supported by DataFrames in Pandas. Here however, we will simply highlight a few ones that might be useful, to give you an idea of what's possible out of the box with Pandas:</p>

##### head()

In [23]:
# can specify the number of rows to be displayed from the header
data.head()

names  ages
0  Alice    29
1    Bob    26
2  James    29
3   Beth    32
4   John    19

##### tail()

In [24]:
# can specify the number of rows to be displayed from the footer
data.tail()


names  ages
5    Sally    32
6  Richard    23
7   Lauren    31
8  Brandon    26
9  Sabrina    31

##### shape

In [ ]:
# The dataframe has a shape property, just like a NumPy matrix. 
# print(df.shape)
print("Shape of DataFrame is: {}".format(data.shape))

# It also has an overall length property corresponding to the number of rows.
print("the DataFrame has a length of: {}".format(len(data)))

##### describe() <br>
<p>Probably one of the most important methods to understand...</p>

In [25]:
# Collect summary statistics in one line
data.describe()

ages
count  10.000000
mean   27.800000
std     4.289522
min    19.000000
25%    26.000000
50%    29.000000
75%    31.000000
max    32.000000

##### sort_values()

In [27]:
# Sort based on many labels, with left-to-right priority
data.sort_values('ages', kind ='mergesort') #inplace defaults to False - have to specify inplace
                                            #to true if you want to actually change the object otherwise 
                                            #its just a copy

names  ages
4     John    19
6  Richard    23
1      Bob    26
8  Brandon    26
0    Alice    29
2    James    29
7   Lauren    31
9  Sabrina    31
3     Beth    32
5    Sally    32

##### .columns

In [28]:
# will show all cols headers
data.columns

Index(['names', 'ages'], dtype='object')

### In-Class Exercise #2 - Describe & Sort Boston Red Sox Hitting Data <br>
<p>Take the data that you read in earlier from the Red Sox csv's and use the describe method to understand the data better. Compare the two years and decide which team is having the better year. Then sort the values based on Batting Average.</p>

In [39]:
print("Boston 2017 Hitting Avg: \n {}".format(boston2017['BA'].describe()['mean']))
print("Boston 2018 Hitting Avg: \n {}".format(boston2018['BA'].describe()['mean']))

mean17 = float((boston2017['BA'].describe()['mean']))
mean18 = float((boston2018['BA'].describe()['mean']))
if( mean17 > mean18):
    print("\n 2017 was a better year")
else:
    print("\n 2018 was a better year")    


Boston 2017 Hitting Avg: 
 0.27843478260869564
Boston 2018 Hitting Avg: 
 0.27843478260869564

 2018 was a better year


### Filtration <br>
<p>Let's look at how to filter dataframes for rows that fulfill a specific conditon.</p>

##### Conditionals

In [ ]:
# Conditional boolean dataframe


##### Subsetting

In [40]:
# exactly like numpy
data[data['ages'] >= 25 ]

names  ages
0    Alice    29
1      Bob    26
2    James    29
3     Beth    32
5    Sally    32
7   Lauren    31
8  Brandon    26
9  Sabrina    31

### Column Transformations <br>
<p>Rarely, if ever, will the columns in the original raw dataframe read from CSV or database table be the ones you actually need for your analysis. You will spend lots of time constantly transforming columns or groups of columns using general computational operations to produce new ones that are functions of the old ones. Pandas has full support for this: Consider the following dataframe containing membership term and renewal number for a group of customers:</p>

In [42]:
# Generate some fake data
customer_id = np.random.randint(1000,1100,10)
renewal_nbr = np.random.randint(0, 10, 10)
customer_dict = { 1: 0.5, 0:1}
term_in_years = [customer_dict[key] for key in np.random.randint(0,2,10)]
# Create a dict - not usually how you would do this
random_data = {
    'customer_id':customer_id,
    'renewal_nbr': renewal_nbr,
    'term_in_years': term_in_years
}
# Build the dataframe
customers = pd.DataFrame.from_dict(random_data)
customers


customer_id  renewal_nbr  term_in_years
0         1056            1            0.5
1         1082            3            0.5
2         1009            0            0.5
3         1022            1            1.0
4         1090            3            1.0
5         1062            4            0.5
6         1011            9            1.0
7         1061            1            0.5
8         1010            6            1.0
9         1076            2            1.0

##### Generating a New Column w/Data

In [45]:
customers['customer_tenure'] = customers['renewal_nbr'] * customers['term_in_years']
customers.head(10)

customer_id  renewal_nbr  term_in_years  customer_tenure
0         1056            1            0.5              0.5
1         1082            3            0.5              1.5
2         1009            0            0.5              0.0
3         1022            1            1.0              1.0
4         1090            3            1.0              3.0
5         1062            4            0.5              2.0
6         1011            9            1.0              9.0
7         1061            1            0.5              0.5
8         1010            6            1.0              6.0
9         1076            2            1.0              2.0

##### User Defined Function

<p>If what you want to do to a column that can't be represented by simple mathematical operations, you can write your own $\textit{user defined function}$ with the full customizability available in Python and any external Python packages, then map it directly onto a column. Let's add some ages to our customer dataframe, and then classify them into our custom defined grouping scheme:</p>

In [47]:
# use .apply to map over dataframe
customers['ages'] = np.random.randint(18,70,10)

def make_age_groups(age):
    if 10 <= age and 20> age:
        return 'Teenager'
    elif 20 <= age and 35 > age:
        return 'Young Adult'
    elif 35 <= age and 65 > age:
        return 'Adult'
    elif 65 <= age:
        return 'Senior'

customers['age_group'] = customers['ages'].apply(make_age_groups)
customers.head(10)


customer_id  renewal_nbr  term_in_years  customer_tenure  ages    age_group
0         1056            1            0.5              0.5    32  Young Adult
1         1082            3            0.5              1.5    68       Senior
2         1009            0            0.5              0.0    46        Adult
3         1022            1            1.0              1.0    48        Adult
4         1090            3            1.0              3.0    55        Adult
5         1062            4            0.5              2.0    19     Teenager
6         1011            9            1.0              9.0    54        Adult
7         1061            1            0.5              0.5    33  Young Adult
8         1010            6            1.0              6.0    56        Adult
9         1076            2            1.0              2.0    50        Adult

<p>As a last example I'll show here how you would use a lambda function to create a UDF that depends on $\textit{more than one}$ column:</p>

<li>UDF = User Defined Function</li>

In [48]:
#Axis for apply can only be 1 or 0 -- 1 being the X axis 0 being the Y axis
def make_loyalty_age_groups(row):
    age = row['ages']
    tenure = row['customer_tenure']
    if 10 <= age and 20> age:
        return 'Teenager'
    elif 20 <= age and 35 > age:
        return 'Young Adult'
    elif 35 <= age and 65 > age:
        return 'Adult'
    elif 65 <= age:
        return 'Senior'
    
    if tenure > 2.0:
        loyalty_age_group = 'Loyal' + age_group
    else:
        loyalty_age_group = 'New' + age_group
    
    return loyalty_age_group

customers['loyalty_age_group'] = customers.apply(make_loyalty_age_groups, axis = 1)

customers.head(10)

customer_id  renewal_nbr  term_in_years  customer_tenure  ages  \
0         1056            1            0.5              0.5    32   
1         1082            3            0.5              1.5    68   
2         1009            0            0.5              0.0    46   
3         1022            1            1.0              1.0    48   
4         1090            3            1.0              3.0    55   
5         1062            4            0.5              2.0    19   
6         1011            9            1.0              9.0    54   
7         1061            1            0.5              0.5    33   
8         1010            6            1.0              6.0    56   
9         1076            2            1.0              2.0    50   

     age_group loyalty_age_group  
0  Young Adult       Young Adult  
1       Senior            Senior  
2        Adult             Adult  
3        Adult             Adult  
4        Adult             Adult  
5     Teenager          Teenager  
6        Adult             Adult  
7  Young Adult       Young Adult  
8        Adult             Adult  
9        Adult             Adult

### In-Class Exercise #3 - Create Your Own UDF <br>
<p>Using the Boston Red Sox data, create your own UDF which creates a new column called 'All-Star' and puts every player with either a batting average over .280 or an on base percentage of over .360 with a result of 'Yes' in the column and 'No' if not.</p>

In [58]:
def makeAllStar(row):
    battingA = row['BA']
    on_base = row['OBP']
    if(battingA > .28 or on_base >.36):
        all_star = 'Yes'
    else:
        all_star = 'No'
    
    return all_star

boston2018['All_Star'] = boston2018.apply(makeAllStar, axis = 1)
boston2018[['Name','BA', 'OBP', 'All_Star']]

    

Name     BA    OBP All_Star
0    Christian Vazquez  0.290  0.330      Yes
1       Mitch Moreland  0.246  0.326       No
2       Dustin Pedroia  0.293  0.369      Yes
3      Xander Bogaerts  0.273  0.343       No
4        Rafael Devers  0.284  0.338      Yes
5    Andrew Benintendi  0.271  0.352       No
6   Jackie Bradley Jr.  0.245  0.323       No
7         Mookie Betts  0.264  0.344       No
8       Hanley Ramirez  0.242  0.320       No
9           Sandy Leon  0.225  0.290       No
10         Chris Young  0.235  0.322       No
11       Deven Marrero  0.211  0.259       No
12       Eduardo Nunez  0.321  0.353      Yes
13          Brock Holt  0.200  0.305       No
14       Josh Rutledge  0.224  0.297       No
15      Pablo Sandoval  0.212  0.269       No
16          Sam Travis  0.263  0.325       No
17         Tzu-Wei Lin  0.268  0.369      Yes
18     Marco Hernandez  0.276  0.300       No
19         Rajai Davis  0.250  0.289       No
20        Steve Selsky  0.111  0.111       No
21       Blake Swihart  0.200  0.429      Yes
22      Chase d'Arnaud  1.000  1.000      Yes

### Aggregations <br>
<p>The raw data plus some transformations is generally only half the story. Your objective is to extract actual insights and actionable conclusions from the data, and that means reducing it from potentially billions of rows to some summary statistics via aggregation functions.</p>

##### groupby() <br>
<p>The .groupby() function is in some ways a 'master' aggregation.</p> 

<p>Data tables will usually reserve one column as a primary key - that is, a column for which each row has a unique value. This is to facilitate access to the exact rows of a data table that a user wants to view. The other columns will often have repeated values, such as the age groups in the above examples. We can use these columns to explore the data using the Pandas API:</p>

In [60]:
# also introducing .count() here, exact same as to how it's used in SQL
customers.groupby('age_group', as_index = False).count()[['age_group', 'customer_id']]

age_group  customer_id
0        Adult            6
1       Senior            1
2     Teenager            1
3  Young Adult            2

##### Type of groupby()

<p>The result is a new dataframe, the columns of which all contain the counts of the grouped field. Notice the type of a grouped dataframe:</p>

In [61]:
type(customers.groupby('age_group', as_index = False))

pandas.core.groupby.generic.DataFrameGroupBy

<p>This is because simply grouping data doesn't quite make sense without an aggregation function like count() to pair with. In this case, we're counting occurances of the grouped field, but that's not all we can do. We can take averages, standard deviations, mins, maxes and much more! Let's see how this works a bit more:</p>

##### mean()

In [62]:
# mean = average
customers.groupby('age_group', as_index=False).mean()[['age_group', 'ages']]

age_group  ages
0        Adult  51.5
1       Senior  68.0
2     Teenager  19.0
3  Young Adult  32.5

##### groupby() w/Multiple Columns

<p>We end up with the average age of the groups in the last column, the average tenure in the tenure column, and so on and so forth. You can even split the groups more finely by passing a list of columns to group by:</p>

In [ ]:
customers.groupby('age_group', as_index=False).mean()[['age_group', 'ages']]

##### drop_duplicates()

<p>Drops all duplicates from the current dataframe</p>

In [64]:
cus_copy = customers

result = cus_copy.drop_duplicates(['age_group'])
result

customer_id  renewal_nbr  term_in_years  customer_tenure  ages  \
0         1056            1            0.5              0.5    32   
1         1082            3            0.5              1.5    68   
2         1009            0            0.5              0.0    46   
5         1062            4            0.5              2.0    19   

     age_group loyalty_age_group  
0  Young Adult       Young Adult  
1       Senior            Senior  
2        Adult             Adult  
5     Teenager          Teenager

<p>Thus the groupby operation allows you to rapidly make summary observations about the state of your entire dataset at flexible granularity. In one line above, we actually did something very complicated - that's the power of the dataframe. In fact, the process often consists of several iterative groupby operations, each revealing greater insight than the last - if you don't know where to start with a dataset, try a bunch of groupbys!</p>

### Homework Excersise #4 - Find the Mean for the Red Sox <br>
<p>For both 2017 and 2018, find the mean of the batting averages</p>

In [7]:
# step 1: Add a new column with the key 'Team' and all column values should be 'BOS'
boston2017['Team'] = 'BOS'
boston2018['Team'] = 'BOS'
# step 2: Group by the 'Team' column and get total home runs and rbi's
print(boston2017.groupby('Team', as_index = False).sum()[['Team', 'HR', 'RBI']])

print(boston2018.groupby('Team', as_index = False).sum()[['Team', 'HR', 'RBI']])


  Team   HR  RBI
0  BOS  168  735
  Team   HR  RBI
0  BOS  208  826
